### PR TITLE
magiclysm: replace spell_effect_line

### PR DESCRIPTION
**Pull request by OrenAudeles**
#### Summary
```SUMMARY: Bugfixes "Fix holes in spell_effect::spell_effect_line"```

#### Purpose of change
Previous implementation occasionally misses tiles that should be hit in some orientations.

#### Describe the solution
Currently does not work outside of 2D space. Allowing 3D line spells is work for the future.
1. Get 2D delta vector
2. Calculate square distance, if zero we can quit early. We are targeting self.
3. Calculate clockwise perpendicular vector from delta vector
4. Find primary axis (north, east, south, west) of delta vector, and from that its clockwise perpendicular axis
5. Calculate clockwise and counter/anti-clockwise axis leg lengths, modify further if distance metric is non-circular
6. Get orientation of delta vector relative to primary axis. This signifies which leg is behind or in front of the clockwise perpendicular axis
7. Get a canonical tile path along delta vector, pop off the endpoint and insert the origin at the beginning. This is done so that we can walk along the path easily.
8. Fill tile path between the source and target points using the line walker
9. Dependant on the orientation found in (6), fill paths for clockwise and counter clockwise legs.

#### Testing
Tossed around some Megablast spells in a group of zombies, did not see holes where there shouldn't be holes.

#### Additional context
All targeting a tile diagonally away from the Player to most easily demonstrate holes in current implementation
- Current Implementation: Without Circular Distance metric
cropped-master-circdist-false
- Current Implementation: With Circular Distance Metric
cropped-master-circdist-true
- PR Implementation: Without Circular Distance Metric
cropped-pr-circdist-false
- PR Implementation: With Circular Distance Metric
cropped-pr-circdist-true

### CODE REVIEW

File: src/magic_spell_effect.cpp [hunk 1]
```diff
@@ -145,78 +145,185 @@ std::set<tripoint> spell_effect::spell_effect_cone( const spell &sp, const tripo
...
+
+    auto test = ignore_walls ?
+    []( const tripoint & ) {
+        return true;
+} :
+    []( const tripoint & p ) {
+        return g->m.passable( p );
+    };
+    // Tests if point c is between or on lines (a0, ao + d) and (a1, a1 + d)
+    auto between_or_on = [side_of]( const point & a0, const point & a1, const point & d,
```
Comment (KorGgenT) [inline src/magic_spell_effect.cpp]:
```suggestion
    auto between_or_on = [side_of]( const point &a0, const point &a1, const point &d,
```

File: src/magic_spell_effect.cpp [hunk 2]
```diff
@@ -145,78 +145,185 @@ std::set<tripoint> spell_effect::spell_effect_cone( const spell &sp, const tripo
...
+    const point & c ) {
```
Comment (KorGgenT) [inline src/magic_spell_effect.cpp]:
```suggestion
    const point &c ) {
```

File: src/magic_spell_effect.cpp [hunk 3]
```diff
@@ -145,78 +145,185 @@ std::set<tripoint> spell_effect::spell_effect_cone( const spell &sp, const tripo
...
+            index = 0;
+        }
+    };
+
+    line_iterable base_line( point_zero, delta, path_to_target );
+
+    std::set<tripoint> result;
+
+    // Builds line until obstructed or outside of region bound by near and far lines
+    auto build_line = [&]( line_iterable line, const tripoint & source, const point & delta,
```
Comment (KorGgenT) [inline src/magic_spell_effect.cpp]:
```suggestion
    auto build_line = [&]( line_iterable line, const tripoint &source, const point &delta,
```

File: src/magic_spell_effect.cpp [hunk 4]
```diff
@@ -145,78 +145,185 @@ std::set<tripoint> spell_effect::spell_effect_cone( const spell &sp, const tripo
...
+    const point & delta_perp ) {
```
Comment (KorGgenT) [inline src/magic_spell_effect.cpp]:
```suggestion
    const point &delta_perp ) {
```

**Comment by OrenAudeles** [comment]
@KorGgenT I think the `blah & name` stuff was the result of astyle being weird with lambda. Issue disappeared when I converted the lambda into functions.

**Comment by KorGgenT** [comment]
I don't have a sufficient enough understanding of what's really going on in this PR to merge it, but it is certainly something i asked for.

**Pull request merged**